src/core/search/compressed_list.cc

```diff
@@ -0,0 +1,210 @@
+  auto diff2_span = WriteVarLen(bound.value - value, absl::MakeSpan(buf2));
+
+  // Extend the location where diff0 is stored with optional zeros before overwriting it
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  size_t required_len = diff1_span.size() + diff2_span.size();
+  DCHECK_LE(bound.diff_span.size(), required_len);  // It can't shrink for sure
+  diffs_.insert(diffs_.begin() + diff_offset, required_len - bound.diff_span.size(), 0u);
+
+  // Now overwrite diff0 and 0s with the two new differences
+  copy(diff1_span.begin(), diff1_span.end(), diffs_.begin() + diff_offset);
+  copy(diff2_span.begin(), diff2_span.end(), diffs_.begin() + diff_offset + diff1_span.size());
+}
+
+// Remove has linear complexity. It tries to find the element V and its neighbors A and B,
+// which are encoded as diff1 = V - A and diff2 = B - V. Adjacently stored diff1 and diff2
+// need to be replaced with diff3 = diff1 + diff2s
+void CompressedList::Remove(uint32_t value) {
+  auto bound = LowerBound(value);
+
+  // Nothing was read or the element was not found
+  if (bound.diff_span.empty() || bound.value != value)
+    return;
+
+  // We're removing below unconditionally
+  size_--;
+
+  // Calculate offset where values diff is stored and determine diffs tail
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  auto diffs_tail = absl::MakeSpan(diffs_).subspan(diff_offset + bound.diff_span.size());
+
+  // If its stored at the end, simply truncate it away
+  if (diffs_tail.empty()) {
+    diffs_.resize(diffs_.size() - bound.diff_span.size());
+    return;
+  }
+
+  // Now the list certainly contains a succeeding element B > V and possibly A < V (or 0)
+  // Read diff2 and calculate diff3 = diff1 + diff2
+  auto [diff2, diff2_read] = ReadVarLen(diffs_tail);
+  uint32_t diff3 = (bound.value - bound.prev_value) + diff2;
+
+  // Encode diff3
+  array<uint8_t, 16> buf;
+  auto diff3_buf = WriteVarLen(diff3, absl::MakeSpan(buf));
+
+  // Shrink vector before overwriting
+  DCHECK_LE(diff3_buf.size(), diff2_read + bound.diff_span.size());
+  size_t to_remove = diff2_read + bound.diff_span.size() - diff3_buf.size();
+  diffs_.erase(diffs_.begin() + diff_offset, diffs_.begin() + diff_offset + to_remove);
+
+  // Overwrite diff1/diff2 with new diff3
+  copy(diff3_buf.begin(), diff3_buf.end(), diffs_.begin() + diff_offset);
+}
+
+size_t CompressedList::Size() const {
+  return size_;
+}
+
+size_t CompressedList::ByteSize() const {
+  return diffs_.size();
+}
+
+// Encode with simple MSB (Most significant bit) encoding: 7 bits of value + 1 to indicate next byte
+absl::Span<uint8_t> CompressedList::WriteVarLen(uint32_t value, absl::Span<uint8_t> buf) {
+  size_t i = 0;
+  do {
+    uint8_t byte = value & 0x7F;
+    value >>= 7;
+    if (value != 0)  // If its not the last byte, set continuation bit
+      byte |= 0x80;
+    buf[i++] = byte;
+  } while (value != 0);
+  return buf.subspan(0, i);
+}
```

Comment:
I didn't find that many interesting solutions and algorithms on the internet. I want it to be more efficient, because running this loop for decoding a single element might be a little slow. 

Sure, we can unroll it manually with ifs. There is also some code in RDB that uses a similar variable length encoding (though its a little more wasteful, it has 1/2/4 bytes gradation IIRC)

Reply:
How about doing something similar to UTF8, in that the header will have a number of `1`s as the size (in bytes) of the serialized data minus 1?
For example, a number that takes 1 byte will be encoded `0.......` (where `.` are data bits), a number taking 2 bytes will be encoded as `10...... ........`, a number taking 3 bytes will be encoded as `110..... ........ ........` etc
(this isn't identical to UTF8, as it spends fewer bits on encoding)
It should "waste" the same number of bits (as non data representing bits), but this way you can tell how many bits to read by examining the bits byte, and also parsing is easier (just zero out the header)
Just a thought..

Reply:
relevant - 
https://lemire.me/blog/2017/09/27/stream-vbyte-breaking-new-speed-records-for-integer-compression/

Reply:
Yes, I'll try this. It should also be possible to implement this without any loops

Reply:
Did you see they have a library with SIMD support? It's licensed Apache 2.0 (which I _think_ we can use)